File: HerrEmil-JavaScript/day-04.js
const parseData = (data) =>
  data.split(/\r?\n\r?\n/).map((passport) =>
    Object.assign(
      ...passport
        .split(/\s+/)
        .map((field) => field.split(":"))
        .map(([key, value]) => ({ [key]: value }))
    )
  );

const requiredFields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"];

const solve1 = (data) =>
  data.filter((passport) =>
    requiredFields.every((field) => passport.hasOwnProperty(field))
  ).length;

const validByr = (byr) => {
  if (!byr) return false;
  const byrNum = parseInt(byr, 10);
  return 1920 <= byrNum && byrNum <= 2002;
};

const validIyr = (byr) => {
  if (!byr) return false;
  const iyrNum = parseInt(byr, 10);
  return 2010 <= iyrNum && iyrNum <= 2020;
};

const validEyr = (byr) => {
  if (!byr) return false;
  const eyrNum = parseInt(byr, 10);
  return 2020 <= eyrNum && eyrNum <= 2030;
};

const validHgt = (hgt) => {
  if (!hgt) return false;
  hgtNum = parseInt(hgt.split(/cm|in/), 10);
  if (hgt.endsWith("cm")) {
    return 150 <= hgtNum && hgtNum <= 193;
  }
  if (hgt.endsWith("in")) {
    return 59 <= hgtNum && hgtNum <= 76;
  }
  return false;
};

const validHcl = (hcl) => /^#[0-9a-f]{6}$/.test(hcl);

const validEcl = (ecl) => {
  if (!ecl) return false;
  return ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"].includes(ecl);
};

const validPid = (pid) => {
  if (!pid) return false;
  return pid.length === 9 && parseInt(pid, 10) !== NaN;
};

const solve2 = (data) =>
  data.filter(({ byr, iyr, eyr, hgt, hcl, ecl, pid }) => {
    if (!validByr(byr)) {
      return false;
    }
    if (!validIyr(iyr)) {
      return false;
    }
    if (!validEyr(eyr)) {
      return false;
    }
    if (!validHgt(hgt)) {
      return false;
    }
    if (!validHcl(hcl)) {
      return false;
    }
    if (!validEcl(ecl)) {
      return false;
    }
    if (!validPid(pid)) {
      return false;
    }
    return true;
  }).length;

module.exports.validByr = validByr;
module.exports.validHgt = validHgt;
module.exports.validHcl = validHcl;
module.exports.validEcl = validEcl;
module.exports.validPid = validPid;
module.exports.solve1 = solve1;
module.exports.solve2 = solve2;
module.exports.parseData = parseData;

File: Dyr-El-python/day04.py
## Start of header boilerplate #################################################

from aocbase import readInput
import re

oneLinePattern = re.compile(r"^.*$")
wsTokenPattern = re.compile(r"\S+")

def lineParse(s, f, fp):
    m = re.findall(fp, s)
    if m==None:
        raise ValueError("No token patterns found")
    return tuple(map(f, m))

def fileParse(inp, lineparser=lineParse,
                   tokenparser=lambda x:x,
                   tokenPattern=re.compile(r"^(.*)$")):
    return tuple(map(lambda x:x, 
                    map(lambda x:lineparser(x, tokenparser, tokenPattern),
                        inp.splitlines())))

## End of header boilerplate ###################################################

def accumulate(itemLines):
    l = list()
    d = dict()
    for line in itemLines:
        if line == ():
            l.append(d)
            d = dict()
        else:
            for item in line:
                a, b = item.split(":")
                d[a] = b
    if len(d) > 0:
        l.append(d)
    return l

def simpleValid(passport):
    return all(item in passport for item in ("byr", "iyr", "eyr",
                                             "hgt", "hcl", "ecl", "pid"))

def part1(pinp):
    return sum(simpleValid(passport) for passport in accumulate(pinp))

def validYear(s, low, high):
    try:
        y = int(s)
    except:
        return False
    return low <= y <= high

def validByr(passport):
    return "byr" in passport and validYear(passport["byr"], 1920, 2002)

def validIyr(passport):
    return "iyr" in passport and validYear(passport["iyr"], 2010, 2020)

def validEyr(passport):
    return "eyr" in passport and validYear(passport["eyr"], 2020, 2030)

def validHgt(passport):
    if "hgt" not in passport:
        return False
    try:
        unit, measure = passport["hgt"][-2:], int(passport["hgt"][:-2])
    except:
        return False
    return ((unit == "cm" and 150 <= measure <= 193) or
            (unit == "in" and 59 <= measure <= 76))

def validHcl(passport):
    if "hcl" not in passport:
        return False
    hcl = passport["hcl"]
    return (len(hcl) == 7 and hcl[0]=="#" and 
            all(c in "0123456789abcdef" for c in hcl[1:]))

def validEcl(passport):
    return ("ecl" in passport and
            passport["ecl"] in ("amb","blu","brn","gry","grn","hzl","oth"))

def validPid(passport):
    if "pid" not in passport:
        return False
    pid = passport["pid"]
    return (len(pid) == 9 and all(c in "0123456789" for c in pid))

def complexValid(passport):
    return (validByr(passport) and validIyr(passport) and validEyr(passport) and
            validHgt(passport) and validHcl(passport) and validEcl(passport) and
            validPid(passport))

def part2(pinp):
    return sum(complexValid(passport) for passport in accumulate(pinp))

## Start of footer boilerplate #################################################

if __name__ == "__main__":
    inp = readInput()
    # inp = """"""
    
    ## Update for input specifics ##############################################
    parseInp = fileParse(inp, tokenPattern=wsTokenPattern)

    print("Input is '" + str(parseInp[:10])[:160] + 
          ('...' if len(parseInp)>10 or len(str(parseInp[:10]))>160 else '') + "'")
    print("Solution to part 1: {}".format(part1(parseInp)))
    print("Solution to part 2: {}".format(part2(parseInp)))

## End of footer boilerplate ###################################################
